### PR TITLE
Add reclient support

### PR DESCRIPTION
**Pull request by goodov**
Add support for remote builds via reclient.

Related devops PR https://github.com/brave/devops/pull/10638

Resolves https://github.com/brave/brave-browser/issues/34030

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: PRESUBMIT.py [hunk 1]
```diff
@@ -286,7 +286,7 @@ def CheckLicense(input_api, output_api):
 
 _BANNED_CPP_FUNCTIONS += (
     BanRule(
-        r'/\bStringPiece',
+        r'/\b(Basic|W)?StringPiece(16)?\b',
```
Comment (bridiver) [inline PRESUBMIT.py:289]: is this relevant to this reclient change?

File: build/commands/lib/config.js [hunk 2]
```diff
@@ -476,6 +484,9 @@ Config.prototype.buildArgs = function () {
     // set goma_dir to the redirect cc output dir which then calls gomacc
     // through env.CC_WRAPPER
     args.goma_dir = path.join(this.nativeRedirectCCDir)
+  } else if (this.useRemoteExec) {
```
Comment (bridiver) [inline build/commands/lib/config.js:487]: should we list this first so it takes precedence over goma settings?

File: build/commands/lib/config.js [hunk 3]
```diff
@@ -1205,8 +1226,23 @@ Object.defineProperty(Config.prototype, 'defaultOptions', {
 
       // Upload stats about Goma actions to the Goma backend.
       env.GOMA_PROVIDE_INFO = true
+    }
+
+    if (this.rbeService) {
```
Comment (bridiver) [inline build/commands/lib/config.js:1231]: what happens if both goma and rbe are set? Maybe as above we should list rbe first and if/else?

File: PRESUBMIT.py (hunk 1 above)
Comment (goodov) [reply PRESUBMIT.py:289]: Yes, it's relevant to `redirect_cc` change in this PR. Without this presubmit complains on `base::FilePath::StringPieceType` usages.

File: build/commands/lib/config.js (hunk 3 above)
Comment (goodov) [reply build/commands/lib/config.js:1231]: > what happens if both goma and rbe are set? Maybe as above we should list rbe first and if/else?

`gn gen` will fail if both goma and remoteexec are enabled. We pass both gn args related to goma and rbe enabling unconditionally.

File: build/commands/lib/config.js (hunk 2 above)
Comment (goodov) [reply build/commands/lib/config.js:487]: > should we list this first so it takes precedence over goma settings?

`gn gen` will fail if both goma and remoteexec are enabled, so the user will be forced to pick one.

Comment (bridiver) [reply build/commands/lib/config.js:487]: Yes, but I'm saying that might not be a particularly desirable failure as even testing this will force people to modify npm config

File: PRESUBMIT.py (hunk 1 above)
Comment (bridiver) [reply PRESUBMIT.py:289]: Yes, but I'm saying that might not be a desirable failure mode because even testing this will require people to delete/modify npm config to switch between them

File: build/commands/lib/config.js (hunk 2 above)
Comment (goodov) [reply build/commands/lib/config.js:487]: > Yes, but I'm saying that might not be a particularly desirable failure as even testing this will force people to modify npm config.

For testing people can do `npm run build -- --use_goma=false --use_remoteexec`. This will override `brave_use_goma=true` set it `.env`.

>  Maybe for now it should default to goma unless you explicitly force reclient on the command line?

You want this to work like so: `--use_goma --use_remoteexec` will use `remoteexec`, but `--use_goma` will just use `goma`? We can do that, but this may silence an incorrect configuration. It will be unclear what exactly is in use currently.

Comment (bridiver) [reply build/commands/lib/config.js:487]: I guess as long as there's some way to override it without changing npm config that's fine

**Review by bridiver — APPROVED** [review]

**Comment by goodov** [comment]
Testing the usual Goma build with changes from this PR before merging to ensure nothing breaks: https://github.com/brave/brave-core/pull/20831
